applications/GeoMechanicsApplication/custom_elements/geo_steady_state_Pw_piping_element.h

```diff
@@ -0,0 +1,297 @@
+        const GeometryType::ShapeFunctionsGradientsType& rShapeFunctionGradients,
+        const Vector&                                    rIntegrationCoefficients) const
+    {
+        RetentionLaw::Parameters RetentionParameters(GetProperties());
+        const auto&              r_properties              = GetProperties();
+        const double             dynamic_viscosity_inverse = 1.0 / r_properties[DYNAMIC_VISCOSITY];
+
+        auto constitutive_matrix = FillPermeabilityMatrix(r_properties[PIPE_HEIGHT]);
+
+        auto result = BoundedMatrix<double, TNumNodes, TNumNodes>{ZeroMatrix{TNumNodes, TNumNodes}};
+        for (unsigned int integration_point_index = 0;
+             integration_point_index < GetGeometry().IntegrationPointsNumber(GetIntegrationMethod());
+             ++integration_point_index) {
+            result += GeoTransportEquationUtilities::CalculatePermeabilityMatrix<TDim, TNumNodes>(
+                rShapeFunctionGradients[integration_point_index], dynamic_viscosity_inverse,
+                constitutive_matrix, 1.0, rIntegrationCoefficients[integration_point_index]);
+        }
+
+        return result;
+    }
+
+    array_1d<double, TNumNodes> GetNodalValuesOf(const Variable<double>& rNodalVariable) const
+    {
+        auto        result     = array_1d<double, TNumNodes>{};
+        const auto& r_geometry = GetGeometry();
+        std::transform(r_geometry.begin(), r_geometry.end(), result.begin(), [&rNodalVariable](const auto& node) {
+            return node.FastGetSolutionStepValue(rNodalVariable);
+        });
+        return result;
+    }
+
+    array_1d<double, TNumNodes> CalculateFluidBodyVector(const Matrix& rNContainer,
+                                                         const GeometryType::ShapeFunctionsGradientsType& rShapeFunctionGradients,
+                                                         const Vector& rIntegrationCoefficients) const
+    {
+        const std::size_t number_integration_points =
+            GetGeometry().IntegrationPointsNumber(GetIntegrationMethod());
+        GeometryType::JacobiansType J_container;
+        J_container.resize(number_integration_points, false);
+        for (std::size_t i = 0; i < number_integration_points; ++i) {
+            J_container[i].resize(GetGeometry().WorkingSpaceDimension(),
+                                  GetGeometry().LocalSpaceDimension(), false);
+        }
+        GetGeometry().Jacobian(J_container, this->GetIntegrationMethod());
+
+        const auto& r_properties        = GetProperties();
+        auto        constitutive_matrix = FillPermeabilityMatrix(r_properties[PIPE_HEIGHT]);
+
+        RetentionLaw::Parameters RetentionParameters(GetProperties());
+
+        array_1d<double, TNumNodes * TDim> volume_acceleration;
+        GeoElementUtilities::GetNodalVariableVector<TDim, TNumNodes>(
+            volume_acceleration, GetGeometry(), VOLUME_ACCELERATION);
+        array_1d<double, TDim> body_acceleration;
```

Comment:
Perhaps we can move this declaration to inside the loop, to keep the variable's scope as limited as possible.

Reply:
done